core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -177,6 +190,41 @@ public RecoveryResponse recoverToTarget() throws IOException {
     }
 
     /**
+     * Determines if the source translog is ready for a sequence-number-based peer recovery. The main condition here is that the source
+     * translog contains all operations between the local checkpoint on the target and the current maximum sequence number on the source.
+     *
+     * @param translogView a view of the translog on the source
+     * @return {@code true} if the source is ready for a sequence-number-based recovery
+     * @throws IOException if an I/O exception occurred reading the translog snapshot
+     */
+    boolean isTranslogReadyForSequenceNumberBasedRecovery(final Translog.View translogView) throws IOException {
+        final long startingSeqNo = request.startingSeqNo();
+        final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
+        logger.trace("starting: [{}], ending: [{}}", startingSeqNo, endingSeqNo);
+        if (startingSeqNo - 1 <= endingSeqNo) {
+            logger.trace(
+                "{} waiting for all operations in the range [{}, {}] to complete",
+                shard.shardId(),
+                startingSeqNo,
+                endingSeqNo);
+            /*
+             * We need to wait for all operations up to the current max to complete, otherwise we can not guarantee that all
+             * operations in the required range will be available for replaying from the translog of the source.
+             */
+            cancellableThreads.execute(() -> shard.waitForOpsToComplete(endingSeqNo));
+
+            final LocalCheckpointTracker tracker = new LocalCheckpointTracker(shard.indexSettings(), startingSeqNo, startingSeqNo - 1);
+            final Translog.Snapshot snapshot = translogView.snapshot();
+            Translog.Operation operation;
+            while ((operation = snapshot.next()) != null) {
+                tracker.markSeqNoAsCompleted(operation.seqNo());
+            }
+            return tracker.getCheckpoint() >= endingSeqNo;
```

Comment:
Put more succinctly: if `startingSeqNo - 1` equals `endingSeqNo` then the local checkpoint on the tracker is initialized to the `endingSeqNo` so the condition and the end of the while loop trivially passes.